File: plugin_files/bin/scripting/0_events.lua
local eventHandlers = {}
local table_unpack = table.unpack

local function CallEventCallbacks(event, eventName, ...)
    for i = 1, #eventHandlers[eventName] do
        local handle = eventHandlers[eventName][i].handle
        if type(handle) == "function" then
            local result = (handle(event, ...) or EventResult.Continue)
            if result ~= EventResult.Continue then return result end
        end
    end

    return EventResult.Continue
end

AddGlobalEvents(function(event, eventName, eventData)
    if not eventHandlers[eventName] then return EventResult.Continue end
    if #eventHandlers[eventName] <= 0 then return EventResult.Continue end

    return CallEventCallbacks(event, eventName, table_unpack(eventData))
end)

AddGlobalEventsJSON(function(event, eventName, eventData)
    if not eventHandlers[eventName] then return EventResult.Continue end
    if #eventHandlers[eventName] <= 0 then return EventResult.Continue end

    return CallEventCallbacks(event, eventName, table_unpack(json.decode(eventData)))
end)

local eventRegistryIndex = 50

function AddEventHandler(eventName, cb)
    if not eventHandlers[eventName] then eventHandlers[eventName] = {} end

    eventRegistryIndex = eventRegistryIndex + 1
    table.insert(eventHandlers[eventName], { eventRegistryIndex = eventRegistryIndex, handle = cb })

    RegisterEventHandlerPlugin(eventName)

    return { key = eventRegistryIndex, name = eventName }
end

function RemoveEventHandler(eventData)
    if not eventData.key or not eventData.name then
        print("Error: Invalid event data passed to RemoveEventHandler.")
        return
    end
    if not eventHandlers[eventData.name] then eventHandlers[eventData.name] = {} end

    for i = 1, #eventHandlers[eventData.name] do
        if eventHandlers[eventData.name].eventRegistryIndex == eventData.key then
            table.remove(eventHandlers[eventData.name], i)
            break
        end
    end

    if #eventHandlers[eventData.name] == 0 then
        RemoveEventHandlerPlugin(eventData.name)
    end
end

function TriggerEvent(eventName, ...)
    local eventResult, eventData, _ = table_unpack(TriggerEventInternal(eventName, json.encode({...})))
    return eventResult, eventData
end

File: plugin_files/bin/scripting/0_events.js
const eventHandlers = {}

const LoadEventFile = (global) => {
    AddGlobalEvents((event, eventName, eventData) => {
        if (!eventHandlers.hasOwnProperty(eventName)) return EventResult.Continue
        if (eventHandlers[eventName].length <= 0) return EventResult.Continue

        eventData.unshift(event);

        const handlers = eventHandlers[eventName]
        for (let i = 0; i < handlers.length; i++) {
            if ((typeof handlers[i].handle) == "function") {
                let result = (handlers[i].handle.apply(null, eventData));
                if (result == null || result == undefined) result = EventResult.Continue;
                if (result != EventResult.Continue) return result
            }
        }

        return EventResult.Continue
    })

    AddGlobalEventsJSON((event, eventName, eventData) => {
        if (!eventHandlers.hasOwnProperty(eventName)) return EventResult.Continue
        if (eventHandlers[eventName].length <= 0) return EventResult.Continue

        eventData = JSON.parse(eventData)
        eventData.unshift(event);

        const handlers = eventHandlers[eventName]
        for (let i = 0; i < handlers.length; i++) {
            if ((typeof handlers[i].handle) == "function") {
                let result = (handlers[i].handle.apply(null, eventData));
                if (result == null || result == undefined) result = EventResult.Continue;
                if (result != EventResult.Continue) return result
            }
        }

        return EventResult.Continue
    })

    let eventRegistryIndex = 50

    global.AddEventHandler = (eventName, cb) => {
        if (!eventHandlers.hasOwnProperty(eventName)) eventHandlers[eventName] = []

        eventRegistryIndex = eventRegistryIndex + 1
        eventHandlers[eventName].push({ eventRegistryIndex, handle: cb })

        RegisterEventHandlerPlugin(eventName)

        return { key: eventRegistryIndex, name: eventName }
    }

    global.RemoveEventHandler = (eventData) => {
        if (typeof (eventData) != "object" || !eventData.key || !eventData.name) {
            console.log("Error: Invalid event data passed to RemoveEventHandler.")
            return
        }
        if (!eventHandlers.hasOwnProperty(eventData.name)) eventHandlers[eventData.name] = []

        for (let i = 0; i < eventHandlers[eventData.name]; i++) {
            if (eventHandlers[eventData.name].eventRegistryIndex == eventData.key) {
                eventHandlers[eventData.name].splice(i, 1)
                break
            }
        }

        if (eventHandlers[eventData.name].length == 0) {
            RemoveEventHandlerPlugin(eventData.name)
        }
    }

    global.TriggerEvent = (eventName, ...args) => {
        return TriggerEventInternal(eventName, JSON.stringify(args))
    }
}

LoadEventFile(globalThis)